### PR TITLE
SRC-6704  Add voice feedback to voice controller

### PR DESCRIPTION
**Pull request by maxzieba**
## Proposed changes

Adding voice feedback for hybrid controller speech control.
User story: https://shadowrobot.atlassian.net/browse/SRC-6704

## Checklist

Before posting a PR ensure that from each of the below categories **AT LEAST ONE BOX HAS BEEN CHECKED**. If more than one category is applicable then more can be checked. Also ensure that the proposed changes have been filled out with relevant information for reviewers.

## Tests

- [X] No tests required to be added. (For small changes that will be tested by CI/CD infrastructure).
- [ ] Added automated tests (if a new class is added (Python or C++), the interface of that class must be unit tested).
- [ ] Manually tested in simulation (if simulation specific or no hardware required to test the functionality). 
- [X] Manually tested on hardware (if hardware specific or related).

## Documentation

- [ ] No documentation required to be added.
- [ ] Added documentation (For any new feature, explain what it does and how to use it. Write the documentation in a relevant space, e.g. Github, Confluence, etc).
- [X] Updated documentation (For changes to pre-existing features mentioned in the documentation).

### CODE REVIEW

File: sr_speech_control/README.md [hunk 1]
```diff
@@ -14,20 +27,23 @@ rosrun sr_speech_control speech_control.py
 The node will permanenly listen to the microphone input, will use Google speech
 recognition to translate audio to text, check if text starts with a chosen trigger word and if so, will publish word after the trigger word to a `sr_speech_control` topic (by default) .
 
-The class used has four parameters that can be provided in order to modify the behaviour:
+The class used has five parameters that can be provided in order to modify the behaviour:
 - `trigger_word` - sets a word that preceeds a command to be sent
-- `command_words` - list of commands that are allowed to be sent
+- `command_words_and_feedback` - dicitonary of commands that are allowed to be sent and corresponding sound feedback words
```
Comment (giusebar) [inline sr_speech_control/README.md]: dictionary

File: sr_speech_control/README.md [hunk 2]
```diff
@@ -14,20 +27,23 @@ rosrun sr_speech_control speech_control.py
...
 - `similar_words_dict_path` - path to a yaml file containing dictionary of words that are easily mistaken for a trigger word or one of the command words
 - `non_speaking_duration` - seconds of non-speaking audio to keep on both sides of the recording
 - `pause_threshold` - seconds of non-speaking audio before a phrase is considered complete
+- `output_device_name` - the device name on which the voice feedback confirmation should be played
```
Comment (giusebar) [inline sr_speech_control/README.md]: we should probably explain how to retrieve the device name. You are inserting logitech, but where does that come from?

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 3]
```diff
@@ -76,9 +105,11 @@ def _recognizer_callback(self, recognizer, audio):
             return
 
         result = [str(x).lower() for x in result.split(' ')]
+        rospy.logwarn("Received: {}".format(result))
```
Comment (giusebar) [inline sr_speech_control/src/sr_speech_control/speech_control.py]: we should change it to rosinfo if we really want to keep this

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 4]
```diff
@@ -76,9 +105,11 @@ def _recognizer_callback(self, recognizer, audio):
...
 
         if self._filter_word(result[0], self.trigger_word) == self.trigger_word:
             command = self._filter_word(''.join(result[1:]), self.command_words)
+            rospy.logwarn("Understand as : {}".format(command))
```
Comment (giusebar) [inline sr_speech_control/src/sr_speech_control/speech_control.py]: same

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 5]
```diff
@@ -91,22 +122,40 @@ def _filter_word(self, word, dictionary, offset=0.5):
...
+    def confirm_voice_command(command, output_device_index=None, sample_rate=None):
+        language = 'en'
+        tts = gTTS(text=command, lang=language, slow=False)
+        fp = BytesIO()
+        tts.write_to_fp(fp)
+        fp.seek(0)
+        audio = AudioSegment.from_file(fp, format="mp3")
+        if output_device_index is not None and sample_rate is not None:
+            audio = audio.set_frame_rate(int(sample_rate))
+        play(audio, output_device_index)
```
Comment (giusebar) [inline sr_speech_control/src/sr_speech_control/speech_control.py:141]: we are still playing if output_device_index is None here. either you add play to the if or set a different device index default

**Review by giusebar — CHANGES_REQUESTED** [review]

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 6]
```diff
@@ -41,6 +49,27 @@ def __init__(self, trigger_word, command_words, command_topic='sr_speech_control
...
+                    self.output_device_sample_rate = p.get_device_info_by_index(index)['defaultSampleRate']
+                    return
+        try:
+            self.output_device_index = p.get_default_output_device_info()['index']
+            self.output_device_sample_rate = p.get_default_output_device_info()['defaultSampleRate']
+        except IOError:
+            device_info = p.get_device_info_by_index(0)
+            self.output_device_index = device_info['index']
+            self.output_device_sample_rate = device_info['defaultSampleRate']
+            rospy.logwarn("No default device found. Using first available output device")
```
Comment (giusebar) [inline sr_speech_control/src/sr_speech_control/speech_control.py]: we should have this function return the device index and device sample rate, and fill the needed variable

File: sr_speech_control/README.md (hunk 1 above)
Comment (maxzieba) [reply sr_speech_control/README.md]: Corrected

File: sr_speech_control/README.md (hunk 2 above)
Comment (maxzieba) [reply sr_speech_control/README.md]: Well, the voice feedback is currently only internally so I avoided printing out once again a complete list of output devices (this happens for the microphone already), so the starting time for this feature is faster. To list all available output devices one could run a quick loop like:

```
p = pyaudio.PyAudio()
for index in range(0, p.get_device_count()):
        rospy.logwarn("{} {}".format(index, p.get_device_info_by_index(index)['name'])
```

I imagine that for the final product of the hybrid controller we will attach a headset and everything will be played on a single device automatically detected or to be selected in the GUI.

If you want I can add a method to select and explain it as Michał did for the microphone

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 5 above)
Comment (maxzieba) [reply sr_speech_control/src/sr_speech_control/speech_control.py:141]: play uses the default audio device when output_device_index is None (https://people.csail.mit.edu/hubert/pyaudio/docs/#class-stream)

File: sr_speech_control/README.md (hunk 2 above)
Comment (maxzieba) [reply sr_speech_control/README.md]: Added a info about suggested device name and how to check for those.

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 4 above)
Comment (maxzieba) [reply sr_speech_control/src/sr_speech_control/speech_control.py]: Sure

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 3 above)
Comment (maxzieba) [reply sr_speech_control/src/sr_speech_control/speech_control.py]: Sure

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 7]
```diff
@@ -41,6 +47,26 @@ def __init__(self, trigger_word, command_words, command_topic='sr_speech_control
...
+
+    @staticmethod
+    def get_output_device_data(device_name=None):
+        p = pyaudio.PyAudio()
+        try:
+            output_device_index = p.get_default_output_device_info()['index']
+            output_device_rate = p.get_default_output_device_info()['defaultSampleRate']
+        except IOError:
+            rospy.logwarn("No default device found. Using first available output device")
+            return None, None
```
Comment (giusebar) [inline sr_speech_control/src/sr_speech_control/speech_control.py]: can you set the variables output_device_index and output_device_rate to None and the beginning and just return them at the end

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 5 above)
Comment (giusebar) [reply sr_speech_control/src/sr_speech_control/speech_control.py:141]: ok

**Review by giusebar — DISMISSED** [review]

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 7 above)
Comment (maxzieba) [reply sr_speech_control/src/sr_speech_control/speech_control.py]: Changed

File: sr_speech_control/src/sr_speech_control/speech_control.py [hunk 8]
```diff
@@ -25,10 +25,24 @@
...
 
+from gtts import gTTS
+from io import BytesIO
+import pyaudio
+
+try:
+    from pydub import AudioSegment
+    from pydub.playback import play
+except (ImportError, ModuleNotFoundError) as e:
+    print("Custom PyDub is not installed. No voice feedback.")
```
Comment (mikramarc) [inline sr_speech_control/src/sr_speech_control/speech_control.py]: please use ros logging

**Review by mikramarc — CHANGES_REQUESTED** [review]

File: sr_speech_control/src/sr_speech_control/speech_control.py (hunk 8 above)
Comment (maxzieba) [reply sr_speech_control/src/sr_speech_control/speech_control.py]: Changed

**Review by mikramarc — APPROVED** [review]

**Review by BenStarmerSmith — APPROVED** [review]

**Pull request merged**